File: SecondYearCS/OOP/CivitasRB/lib/vista_textual.rb
#encoding:utf-8
require_relative 'operaciones_juego'
require_relative 'respuestas.rb'
require_relative 'gestiones_inmobiliarias.rb'
require_relative 'salidas_carcel.rb'
require_relative 'civitas_juego.rb'
require 'io/console'
module CivitasRB
  class VistaTextual
    
    def initialize()
      @iGestion = -1
      @iPropiedad = -1
      @juegoModel = nil
      @Respuestas = [Respuestas::SI, Respuestas::NO]
      @GestionesInmobiliarias = [GestionesInmobiliarias::VENDER,
        GestionesInmobiliarias::HIPOTECAR, GestionesInmobiliarias::CANCELAR_HIPOTECA,
        GestionesInmobiliarias::CONSTRUIR_CASA, GestionesInmobiliarias::CONSTRUIR_HOTEL,
        GestionesInmobiliarias::TERMINAR
      ]
      @SalidasCarcel = [SalidasCarcel::PAGANDO,SalidasCarcel::TIRANDO]
    end

    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end
    
    def salirCarcel
      arr = ["Pagando","Tirando"]
      opcion = menu("Elige la forma para intentar salir de la carcel",
      arr)
    return @SalidasCarcel[opcion]
    end

    def comprar
      arr = ["Si", "No"]
      opcion = menu("Deseas comprar la Propiedad?", arr)
      return @Respuestas[opcion]
    end

    def gestionar
      arr = ["Vender","Hipotecar","Cancelar_hipoteca","Construir_Casa","Construir_Hotel","Terminar"]
      opcion = menu("Deseas Gestionar su propiedad?", arr)
      @iGestion = opcion
      lista = @juegoModel.getJugador().getPropiedades()
      lista.each_with_index do |titulo,index|
        if titulo == @juegoModel.getCasillaActual.getTituloPropiedad()
          @iPropiedad = index
        end
      end
    end

    def getGestion
      return @iGestion
    end

    def getPropiedad
      return @iPropiedad
    end

    def mostrarSiguienteOperacion(operacion)
      self.mostrar_estado(operacion.to_s)
    end

    def mostrarEventos
      while(Diario.instance.eventos_pendientes)
        self.mostrar_estado(Diario.instance.leer_evento)
      end
    end

    def setCivitasJuego(civitas)
         @juegoModel=civitas
         self.actualizarVista
    end

    def actualizarVista
      puts "----------------------------------------"
      puts @juegoModel.getJugador().to_s
      puts @juegoModel.getCasillaActual.to_s
      puts "----------------------------------------"
    end

  end
end

File: SecondYearCS/OOP/Civitas/src/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar() {
       int opcion = menu ("Deseas comprar la Propiedad?",
      new ArrayList<> (Arrays.asList("Si","No")));
    return (Respuestas.values()[opcion]);
  }

  void gestionar () {
      int opcion = menu ("Deseas Gestionar su propiedad?",
      new ArrayList<> (Arrays.asList("Vender","Hipotecar",
      "CancelarHipoteca","ConstruirCasa","ConstruirHotel","Terminar")));
      this.iGestion = opcion;
      ArrayList<TituloPropiedad> lista =  juegoModel.getJugador().getPropiedades();
      for(int i=0;i<lista.size();i++){
          if(lista.get(i) == juegoModel.getCasillaActual().getTituloPropiedad());
              this.iPropiedad = i;
      }
          
  }
  
  public int getGestion(){
      return this.iGestion;
  }
  
  public int getPropiedad(){
      return this.iPropiedad;
  }
    

  void mostrarSiguienteOperacion(OperacionesJuego operacion) {
      this.mostrarEstado(operacion.toString());
  }


  void mostrarEventos() {
      while(Diario.getInstance().eventosPendientes()){
          this.mostrarEstado(Diario.getInstance().leerEvento());
      }
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();
    }
  
  void actualizarVista(){
      System.out.println("#########################################");
      System.out.println(juegoModel.getJugador().toString());
      System.out.println(juegoModel.getCasillaActual().toString());
      System.out.println("#########################################");

  } 
}
